cmd/serve/serve.go

```diff
@@ -33,32 +33,30 @@ import (
 	"github.com/cilium/cilium/pkg/monitor/agent/listener"
 	monitorAPI "github.com/cilium/cilium/pkg/monitor/api"
 	"github.com/cilium/cilium/pkg/monitor/payload"
+	"github.com/gogo/protobuf/types"
+	"github.com/google/gops/agent"
+	"github.com/spf13/cobra"
+	"go.uber.org/zap"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/health"
+	healthpb "google.golang.org/grpc/health/grpc_health_v1"
+
 	pb "github.com/cilium/hubble/api/v1/flow"
```

Comment:
did you do this by hand or an editor of yours did this?

Reply:
gofmt

Reply:
gofmt doesn't reformat import sections. It looks like maybe you have a custom goimports config with local variable?

By default goimports without any parameters just does two sections:

```
> head -40 cmd/serve/serve.go
package serve

import (
        "bytes"
        "encoding/gob"
        "fmt"
        "io"
        "net"
        "net/http"
        _ "net/http/pprof" // a comment justifying it
        "os"
        "os/signal"
        "strings"
        "syscall"
        "time"
        "github.com/cilium/cilium/pkg/defaults"
        "github.com/cilium/cilium/pkg/monitor"
        "github.com/cilium/cilium/pkg/monitor/agent/listener"
        monitorAPI "github.com/cilium/cilium/pkg/monitor/api"
        "github.com/cilium/cilium/pkg/monitor/payload"
        pb "github.com/cilium/hubble/api/v1/flow"
        "github.com/cilium/hubble/api/v1/observer"
        "github.com/cilium/hubble/pkg/api"
        v1 "github.com/cilium/hubble/pkg/api/v1"
        "github.com/cilium/hubble/pkg/cilium/client"
        "github.com/cilium/hubble/pkg/fqdncache"
        "github.com/cilium/hubble/pkg/ipcache"
        "github.com/cilium/hubble/pkg/metrics"
        metricsAPI "github.com/cilium/hubble/pkg/metrics/api"
        "github.com/cilium/hubble/pkg/parser"
        "github.com/cilium/hubble/pkg/server"
        "github.com/cilium/hubble/pkg/servicecache"
        "github.com/gogo/protobuf/types"
        "github.com/google/gops/agent"
        "github.com/spf13/cobra"
        "go.uber.org/zap"
        "google.golang.org/grpc"
        "google.golang.org/grpc/health"
        healthpb "google.golang.org/grpc/health/grpc_health_v1"
)

> goimports cmd/serve/serve.go | head -40
package serve

import (
        "bytes"
        "encoding/gob"
        "fmt"
        "io"
        "net"
        "net/http"
        _ "net/http/pprof" // a comment justifying it
        "os"
        "os/signal"
        "strings"
        "syscall"
        "time"

        "github.com/cilium/cilium/pkg/defaults"
        "github.com/cilium/cilium/pkg/monitor"
        "github.com/cilium/cilium/pkg/monitor/agent/listener"
        monitorAPI "github.com/cilium/cilium/pkg/monitor/api"
        "github.com/cilium/cilium/pkg/monitor/payload"
        pb "github.com/cilium/hubble/api/v1/flow"
        "github.com/cilium/hubble/api/v1/observer"
        "github.com/cilium/hubble/pkg/api"
        v1 "github.com/cilium/hubble/pkg/api/v1"
        "github.com/cilium/hubble/pkg/cilium/client"
        "github.com/cilium/hubble/pkg/fqdncache"
        "github.com/cilium/hubble/pkg/ipcache"
        "github.com/cilium/hubble/pkg/metrics"
        metricsAPI "github.com/cilium/hubble/pkg/metrics/api"
        "github.com/cilium/hubble/pkg/parser"
        "github.com/cilium/hubble/pkg/server"
        "github.com/cilium/hubble/pkg/servicecache"
        "github.com/gogo/protobuf/types"
        "github.com/google/gops/agent"
        "github.com/spf13/cobra"
        "go.uber.org/zap"
        "google.golang.org/grpc"
        "google.golang.org/grpc/health"
        healthpb "google.golang.org/grpc/health/grpc_health_v1"

```

You can add `-local` to the goimports parameters to sort local imports into the third bucket (which we haven't really been doing recently).

```
> goimports -local github.com/cilium/hubble cmd/serve/serve.go | head -40
package serve

import (
        "bytes"
        "encoding/gob"
        "fmt"
        "io"
        "net"
        "net/http"
        _ "net/http/pprof" // a comment justifying it
        "os"
        "os/signal"
        "strings"
        "syscall"
        "time"

        "github.com/cilium/cilium/pkg/defaults"
        "github.com/cilium/cilium/pkg/monitor"
        "github.com/cilium/cilium/pkg/monitor/agent/listener"
        monitorAPI "github.com/cilium/cilium/pkg/monitor/api"
        "github.com/cilium/cilium/pkg/monitor/payload"
        "github.com/gogo/protobuf/types"
        "github.com/google/gops/agent"
        "github.com/spf13/cobra"
        "go.uber.org/zap"
        "google.golang.org/grpc"
        "google.golang.org/grpc/health"
        healthpb "google.golang.org/grpc/health/grpc_health_v1"

        pb "github.com/cilium/hubble/api/v1/flow"
        "github.com/cilium/hubble/api/v1/observer"
        "github.com/cilium/hubble/pkg/api"
        v1 "github.com/cilium/hubble/pkg/api/v1"
        "github.com/cilium/hubble/pkg/cilium/client"
        "github.com/cilium/hubble/pkg/fqdncache"
        "github.com/cilium/hubble/pkg/ipcache"
        "github.com/cilium/hubble/pkg/metrics"
        metricsAPI "github.com/cilium/hubble/pkg/metrics/api"
```

It's not really written down anywhere as a rule. I will reconfigure mine to use three groups as well.

Reply:
yeah sorry i meant goimports, i'm using it with goland file watcher. i added -local, vaguely remembering you and @gandro talking about the import order at some point.